crates/circuit/src/circuit_data.rs

```diff
@@ -85,41 +92,261 @@ pub struct CircuitData {
+            param_table: ParamTable::new(),
+            global_phase,
+        };
+        if num_qubits > 0 {
+            let qubit_cls = QUBIT.get_bound(py);
+            for _i in 0..num_qubits {
+                let bit = qubit_cls.call0()?;
+                res.add_qubit(py, &bit, true)?;
+            }
+        }
+        for (operation, params, qargs) in instruction_iter {
+            let qubits = PyTuple::new_bound(py, res.qubits.map_indices(&qargs)).unbind();
+            let clbits = PyTuple::empty_bound(py).unbind();
+            let inst = res.pack_owned(
+                py,
+                &CircuitInstruction {
+                    operation: OperationType::Standard(operation),
+                    qubits,
+                    clbits,
+                    params,
+                    extra_attrs: None,
+                    #[cfg(feature = "cache_pygates")]
+                    py_op: None,
+                },
+            )?;
+            res.data.push(inst);
+        }
+        Ok(res)
     }
-}
 
-impl From<CacheFullError> for PyErr {
-    fn from(_: CacheFullError) -> Self {
-        PyRuntimeError::new_err("The bit operands cache is full!")
+    fn handle_manual_params(
+        &mut self,
+        py: Python,
+        inst_index: usize,
+        params: &[(usize, Vec<PyObject>)],
+    ) -> PyResult<bool> {
+        let mut new_param = false;
+        let mut atomic_parameters: HashMap<u128, PyObject> = HashMap::new();
+        for (param_index, raw_param_objs) in params {
+            raw_param_objs.iter().for_each(|x| {
+                atomic_parameters.insert(
+                    x.getattr(py, intern!(py, "_uuid"))
+                        .expect("Not a parameter")
+                        .getattr(py, intern!(py, "int"))
+                        .expect("Not a uuid")
+                        .extract::<u128>(py)
+                        .unwrap(),
+                    x.clone_ref(py),
+                );
+            });
+            for (param_uuid, param_obj) in atomic_parameters.iter() {
+                match self.param_table.table.get_mut(param_uuid) {
+                    Some(entry) => entry.add(inst_index, *param_index),
+                    None => {
+                        new_param = true;
+                        let new_entry = ParamEntry::new(inst_index, *param_index);
+                        self.param_table
+                            .insert(py, param_obj.clone_ref(py), new_entry)?;
+                    }
+                };
+            }
+            atomic_parameters.clear()
+        }
+        Ok(new_param)
+    }
+
+    /// Add an instruction's entries to the parameter table
+    fn update_param_table(
+        &mut self,
+        py: Python,
+        inst_index: usize,
+        params: Option<Vec<(usize, Vec<PyObject>)>>,
+    ) -> PyResult<bool> {
+        if let Some(params) = params {
+            return self.handle_manual_params(py, inst_index, &params);
+        }
```

Comment:
What I meant about "`handle_manual_params` should be a separate method" is that we should remove the `params` argument from this function, and just have callers directly call your new `handle_manual_params`.